benchmarks/prescribe_dike_injection/prescribed_dike_injection.cc

```diff
@@ -0,0 +1,571 @@
+                      composition_evaluators[c]->evaluate({old_solution_values.data(),old_solution_values.size()},
+                                                          EvaluationFlags::values);
+                      const double old_solution_other_composition = composition_evaluators[c]->get_value(0);
+
+                      // When new dike material is injected, the other compositional fields
+                      // at the dike point will be reduced in the same proportion (p_c) to
+                      // ensure that the sum of all compositional fields is always 1.0.
+                      // For example, in the previous step, the dike material has a compostional
+                      // field with the value of c_dike_old and another compositional field
+                      // with a value of c_1_old. So, c_dike_old + c_1_old = 1.0. Here we
+                      // leave the background field alone, because it will be automatically
+                      // populated if c_dike_old + c_1_old < 1.0.
+                      // In the currest step, when adding a new dike material of amount 'c_dike_add',
+                      // c_dike_new =  c_dike_old + c_dike_add. c_1_new = c_1_old * p_c.
+                      // Since c_1_new + c_dike_new = 1.0, we get 
+                      // p_c = (1.0 - c_dike_old - c_dike_add) / c_1_old.
+                      // Then the amount of change in c_1 is:
+                      // delta_c_1 = c_1_new - c_1_old = c_1_old * (p_c - 1.0)
+                      // = - c_1_old * (c_dike_add / (1.0001 - c_dike_old))
+                      // To avoid dividing by 0, we will use 1.0001 instead of 1.0.
+
+                      // We limit the value of injection phase compostional
+                      // field at previous timestep is [0,1].
+                      double injection_phase_composition = std::max(std::min(composition_evaluators[injection_phase_index]->get_value(0),1.0),0.0);
+
+                      out.reaction_terms[i][c] = -old_solution_other_composition 
+                                                 * std::min(dike_injection_fraction 
+                                                 / (1.0001 - injection_phase_composition), 1.0);
+
+                      // Fill reaction rate outputs instead of the reaction terms if
+                      // we use operator splitting (and then set the latter to zero).
+                      if (reaction_rate_out != nullptr)
+                        reaction_rate_out->reaction_rates[i][c] = out.reaction_terms[i][c]
+                                                                  / this->get_timestep();
+
+                      if (this->get_parameters().use_operator_splitting)
+                        out.reaction_terms[i][c] = 0.0;
+                    }
+                }
+
+              // If the "single Advection" nonlinear solver scheme is used, 
+              // it is necessary to set the reaction term to 0 to avoid
```

Comment:
Because I remember that in some of my models the effect of dilation on the strain rate was not fully corrected, especially at the boundaries of the dikes. I guess the reason is that even if I set the tolerance of the nonlinear solver to 1e-4, the convergence is not good, resulting in a residual effect after the correction, which leads to an unexpected increase in the plastic strain. Therefore, here I simply set the increase in plastic strain of the dike to 0.